lang/rust/avro/src/encode.rs

```diff
@@ -45,155 +50,177 @@ fn encode_int(i: i32, buffer: &mut Vec<u8>) {
     zig_i32(i, buffer)
 }
 
-/// Encode a `Value` into avro format.
-///
-/// **NOTE** This will not perform schema validation. The value is assumed to
-/// be valid with regards to the schema. Schema are needed only to guide the
-/// encoding for complex type values.
-pub fn encode_ref(value: &Value, schema: &Schema, buffer: &mut Vec<u8>) {
-    fn encode_ref0(
-        value: &Value,
-        schema: &Schema,
-        buffer: &mut Vec<u8>,
-        schemas_by_name: &mut HashMap<Name, Schema>,
-    ) {
-        match &schema {
-            Schema::Ref { ref name } => {
-                let resolved = schemas_by_name.get(name).unwrap();
-                return encode_ref0(value, resolved, buffer, &mut schemas_by_name.clone());
-            }
-            Schema::Record { ref name, .. }
-            | Schema::Enum { ref name, .. }
-            | Schema::Fixed { ref name, .. } => {
-                schemas_by_name.insert(name.clone(), schema.clone());
-            }
-            _ => (),
-        }
+fn encode_internal(
+    value: &Value,
+    schema: &Schema,
+    names: &HashMap<Name, &Schema>,
+    enclosing_namespace: &Namespace,
+    buffer: &mut Vec<u8>,
+) -> AvroResult<()> {
+    if let Schema::Ref { ref name } = schema {
+        let resolved = *names
+            .get(&name.fully_qualified_name(enclosing_namespace))
```

Comment:
`name.fully_qualified_name(enclosing_namespace)` could be extracted into a local variable and re-used at line 64